### PR TITLE
Some more protocol handler refactoring. WORK IN PROGRESS

### PR DESCRIPTION
**Pull request by andre4i**
This PR cannot build and should not be merged, as it spans monorepos (in the context of lerna) and versions and it was/will be tested using `--symlink:full`. It will be split after testing and incorporating feedback.

What this does:

- moves audience management into the protocol handler (along with the IAudience interface into protocol definitions)
- adds a protocol handler builder type
- all public entry points into the container has an optional `IProtocolDetails` parameter which can be used to inject a custom protocol handler

~The PR is also waiting on some changes from `main` for the `Container` class. Until then, they have been brought in manually.~

### CODE REVIEW

File: common/lib/protocol-definitions/api-report/protocol-definitions.api.md [hunk 1]
```diff
@@ -40,6 +41,17 @@ export interface IAttachment {
     id: string;
 }
 
+// @public
+export interface IAudience extends EventEmitter {
```
Comment (andre4i) [inline common/lib/protocol-definitions/api-report/protocol-definitions.api.md:45]: I more than open to suggestions about how to deal with this. I'm adding it to the protocol-definitions as it's the lowest common ancestor of all packages who would need it, in the context of NOT adding a dependency between container definitions and protocol base (which should not be done IMO).

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -544,7 +570,7 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
                 },
                 error);
             });
-        this._audience = new Audience();
+        this._audience = audience ?? new Audience();
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:573]: If we go with this route, I'd suggest to creating Audience same way as Quorum is created - i.e. when protocol is initialized.
I think the reason it's done the way it's done - handling of "connect" event and operating on audience long before protocol is setup. I'd suggest changing it by synthetizing fake signals and pushing them into  this._deltaManager.inboundSignal. That way all processing and all initialization is symmetrical.

Reason I like it (independently from this PR) - people get really confused about audience and quorum. I think we own them revisiting this area, and ideally (if it's possible) - expose single data structure to track users in document. At least at Container (host interaction level), where we have a bit more room to "lie" by exposing user (clientId) presence in simpler way. Story for DDSs is different - there we rely on specific order of quorum operations (all consensus data structures), so we have to expose conceptually data roughly in same way. But that's unknown (maybe thing never to happen), so maybe we do not care?

File: common/lib/protocol-definitions/src/consensus.ts [hunk 3]
```diff
@@ -98,3 +99,36 @@ export interface IProtocolState {
 export interface IProcessMessageResult {
     immediateNoOp?: boolean;
 }
+
+/**
+ * Audience represents all clients connected to the op stream, both read-only and read/write.
+ */
+ export interface IAudience extends EventEmitter {
```
[the remark this answers is not in the document]
Comment (andre4i) [reply common/lib/protocol-definitions/src/consensus.ts:106]: `protocol-base` needs it, so that would imply adding a dependency to `container-definitions`, which is under `packages/common`, not `common/lib`. Haven't tried it yet, but I don't think that dependency is viable.

File: common/lib/protocol-definitions/api-report/protocol-definitions.api.md (hunk 1 above)
Comment (andre4i) [reply common/lib/protocol-definitions/api-report/protocol-definitions.api.md]: This is a non-issue now, this can be left here, as the custom protocol handler which needs an audience has been brought into the loader for now

File: common/lib/protocol-definitions/src/consensus.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (andre4i) [reply common/lib/protocol-definitions/src/consensus.ts]: No longer an issue https://github.com/microsoft/FluidFramework/pull/10818#discussion_r910409346

File: packages/loader/container-loader/src/container.ts (hunk 2 above)
Comment (andre4i) [reply packages/loader/container-loader/src/container.ts]: I'm wary of creating and sending signals in that part of the code. Also, the interface provided by the delta manager for it's signal queue is limited to reading so I'd have to open up the delta manager to allow for this.. I have opted for storing the existing clients explicitly and providing them to the builder, and at the same time add them to the audience iff the protocol handler is already instantiated. I believe it will have the same effect.

**Comment by andre4i** [comment]
Hi @vladsud, this has been updated with:

- moving container-friendly protocol implementations outside the server monorepo
- no more protocol-definitions changes
- audience/protocol handler are created at the same time by storing the existing clients and replaying them at audience creation
- ILoaderProps has the ability to override the default protocol - need to think of an elegant way to make some tests run with empty quorums and such consistently

Take a look when you have some time

**Comment by andre4i** [comment]
@zagriswo FYI

File: packages/common/container-definitions/src/audience.ts [hunk 4]
```diff
@@ -28,4 +28,20 @@ export interface IAudience extends EventEmitter {
      * or undefined if the specified client isn't connected
      */
     getMember(clientId: string): IClient | undefined;
+
```
Comment (vladsud) [inline packages/common/container-definitions/src/audience.ts:31]: I believe we have to separate this interface into two - one that is available to most parties (including runtime itself), and one - that is available to Container to manipulate it. Container runtime should not be in position to call such APIs as clear()

File: packages/loader/container-loader/src/container.ts [hunk 5]
```diff
@@ -330,16 +335,21 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
     public static async rehydrateDetachedFromSnapshot(
         loader: Loader,
         snapshot: string,
+        protocolDetails?: IProtocolDetails,
     ): Promise<Container> {
         const container = new Container(
             loader,
             {});
+
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:343]: I wonder - should we instead pass protocolDetails as constructor argument? And thus not need to change various pass-through functions?

File: packages/loader/container-loader/src/container.ts [hunk 6]
```diff
@@ -1350,28 +1378,30 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
-        const protocol = new ProtocolOpHandlerWithClientValidation(
-            attributes.minimumSequenceNumber,
-            attributes.sequenceNumber,
-            attributes.term,
-            members,
-            proposals,
-            values,
-            (key, value) => this.submitMessage(MessageType.Propose, { key, value }),
-        );
+        const protocol = protocolHandlerBuilder !== undefined ?
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1392]: Feels like code would be more readable if function argument would have default value, i.e.
```ts
protocolHandlerBuilder: ProtocolHandlerBuilder = (...args) => new ProtocolHandler(...args),
```

File: packages/loader/container-loader/src/container.ts [hunk 7]
```diff
@@ -1493,12 +1523,16 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
 
-        deltaManager.on("connect", (details: IConnectionDetails, opsBehind?: number) => {
-            // Back-compat for new client and old server.
-            this._audience.clear();
+        deltaManager.on("connect", (details: IConnectionDetails, _opsBehind?: number) => {
+            this._initialClients = details.initialClients ?? [];
 
-            for (const priorClient of details.initialClients ?? []) {
-                this._audience.addMember(priorClient.clientId, priorClient.client);
+            if (this._protocolHandler !== undefined) {
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1529]: I think it will not be Ok to drop initial clients on the floor if we do not have connection. This will not impact 1st party scenarios (that all delay connection until after boot), but default mode runs connection in parallel with rest of boot, and this will miss initial audience population.

Comment (andre4i) [reply packages/loader/container-loader/src/container.ts:1529]: They aren't dropped, they're saved inside `this._initialClients` (see 1527) and then fed into the protocol handler when it's created.

File: packages/loader/container-loader/src/container.ts [hunk 8]
```diff
@@ -1493,12 +1523,16 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
+                // Back-compat for new client and old server.
```
[the remark this answers is not in the document]
Comment (andre4i) [reply packages/loader/container-loader/src/container.ts:1530]: this was copy-pasted, I'll remove it.

File: packages/loader/container-loader/src/container.ts (hunk 5 above)
Comment (andre4i) [reply packages/loader/container-loader/src/container.ts:343]: will do.

File: packages/loader/container-loader/src/container.ts (hunk 7 above)
Comment (andre4i) [reply packages/loader/container-loader/src/container.ts:1529]: The `if` statement here is to ensure that in case of a reconnection, these are fed into the protocolhandler as soon as the reconnection happens.

Comment (vladsud) [reply packages/loader/container-loader/src/container.ts:1539]: Got it! Maybe make a comment here about it - it's not obvious.

**Comment by andre4i** [comment]
This needs to go to `main` now

**Comment by andre4i** [comment]
Moved to https://github.com/microsoft/FluidFramework/pull/11109
